File: algorithms/python/algorithms.py
#!/usr/bin/env python3
"""
This file is part of the Python Algorithms project.
Sorting algorithms implementation with performance measurement and JSON output.
"""

import json
import time
import argparse
import sys
import os
import threading
import statistics
from typing import List, Tuple, Callable, Dict
from concurrent.futures import ThreadPoolExecutor
import random
import math


# Global mutex for thread-safe JSON results writing
results_lock = threading.Lock()


# --- Sorting algorithms ---

def bubble_sort(array: List[int]) -> None:
    """Bubble sort implementation"""
    size = len(array)
    
    if size < 2:
        return
    
    while True:
        swapped = False
        
        for index in range(1, size):
            if array[index - 1] > array[index]:
                array[index - 1], array[index] = array[index], array[index - 1]
                swapped = True
        
        size -= 1
        if not swapped:
            break


def selection_sort(array: List[int]) -> None:
    """Selection sort implementation"""
    size = len(array)
    
    if size < 2:
        return
    
    for index in range(size - 1):
        min_index = index
        
        for j in range(index + 1, size):
            if array[j] < array[min_index]:
                min_index = j
        
        array[index], array[min_index] = array[min_index], array[index]


def insertion_sort(array: List[int]) -> None:
    """Insertion sort implementation"""
    size = len(array)
    
    if size < 2:
        return
    
    for index in range(1, size):
        key = array[index]
        j = index - 1
        
        while j >= 0 and array[j] > key:
            array[j + 1] = array[j]
            j -= 1
        
        array[j + 1] = key


def quick_sort(array: List[int]) -> None:
    """Quick sort implementation"""
    if len(array) < 2:
        return
    
    def partition(arr: List[int], low: int, high: int) -> int:
        pivot = arr[high]
        i = low - 1
        
        for j in range(low, high):
            if arr[j] < pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        return i + 1
    
    def quick_sort_recursive(arr: List[int], low: int, high: int) -> None:
        if low < high:
            pi = partition(arr, low, high)
            quick_sort_recursive(arr, low, pi - 1)
            quick_sort_recursive(arr, pi + 1, high)
    
    quick_sort_recursive(array, 0, len(array) - 1)


def merge_sort(array: List[int]) -> None:
    """Merge sort implementation"""
    if len(array) < 2:
        return
    
    def merge(arr: List[int], left: int, mid: int, right: int) -> None:
        n1 = mid - left + 1
        n2 = right - mid
        
        L = arr[left:mid + 1]
        R = arr[mid + 1:right + 1]
        
        i = j = 0
        k = left
        
        while i < n1 and j < n2:
            if L[i] <= R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1
        
        while i < n1:
            arr[k] = L[i]
            i += 1
            k += 1
        
        while j < n2:
            arr[k] = R[j]
            j += 1
            k += 1
    
    def merge_sort_recursive(arr: List[int], left: int, right: int) -> None:
        if left < right:
            mid = left + (right - left) // 2
            merge_sort_recursive(arr, left, mid)
            merge_sort_recursive(arr, mid + 1, right)
            merge(arr, left, mid, right)
    
    merge_sort_recursive(array, 0, len(array) - 1)


def heap_sort(array: List[int]) -> None:
    """Heap sort implementation"""
    size = len(array)
    
    if size < 2:
        return
    
    def heapify(arr: List[int], n: int, i: int) -> None:
        largest = i
        left = 2 * i + 1
        right = 2 * i + 2
        
        if left < n and arr[left] > arr[largest]:
            largest = left
        
        if right < n and arr[right] > arr[largest]:
            largest = right
        
        if largest != i:
            arr[i], arr[largest] = arr[largest], arr[i]
            heapify(arr, n, largest)
    
    # Build max heap
    for i in range(size // 2 - 1, -1, -1):
        heapify(array, size, i)
    
    # Extract elements from heap one by one
    for i in range(size - 1, 0, -1):
        array[0], array[i] = array[i], array[0]
        heapify(array, i, 0)


def counting_sort(array: List[int]) -> None:
    """Counting sort implementation"""
    if len(array) < 2:
        return
    
    max_value = max(array)
    min_value = min(array)
    range_val = max_value - min_value + 1
    count = [0] * range_val
    
    for num in array:
        count[num - min_value] += 1
    
    index = 0
    for i in range(range_val):
        while count[i] > 0:
            array[index] = i + min_value
            index += 1
            count[i] -= 1


def radix_sort(array: List[int]) -> None:
    """Radix sort implementation"""
    if len(array) < 2:
        return
    
    max_value = max(array)
    exp = 1
    
    while max_value // exp > 0:
        output = [0] * len(array)
        count = [0] * 10
        
        for num in array:
            count[(num // exp) % 10] += 1
        
        for i in range(1, 10):
            count[i] += count[i - 1]
        
        for i in range(len(array) - 1, -1, -1):
            output[count[(array[i] // exp) % 10] - 1] = array[i]
            count[(array[i] // exp) % 10] -= 1
        
        for i in range(len(array)):
            array[i] = output[i]
        
        exp *= 10


def bucket_sort(array: List[int]) -> None:
    """Bucket sort implementation"""
    if len(array) < 2:
        return
    
    max_value = max(array)
    bucket_count = int(math.sqrt(len(array)))
    buckets = [[] for _ in range(bucket_count)]
    
    for num in array:
        bucket_index = min(int((num / (max_value + 1)) * bucket_count), bucket_count - 1)
        buckets[bucket_index].append(num)
    
    array.clear()
    
    for bucket in buckets:
        if len(bucket) < 2:
            array.extend(bucket)
        else:
            bucket.sort()
            array.extend(bucket)


# --- Utility functions ---

def read_file(file_path: str) -> List[int]:
    """Read integers from a file"""
    try:
        with open(file_path, 'r') as file:
            content = file.read().strip()
            if not content:
                return []
            return [int(x) for x in content.split()]
    except FileNotFoundError:
        raise FileNotFoundError(f"Could not open file: {file_path}")
    except ValueError as e:
        raise ValueError(f"Invalid data in file: {e}")


def run_sort_multiple(algorithm: str, sort_function: Callable[[List[int]], None], 
                     data: List[int], runs: int = 10) -> Tuple[List[float], str]:
    """Run a sorting algorithm multiple times and return timing results"""
    times = []
    
    for _ in range(runs):
        data_copy = data.copy()
        start_time = time.perf_counter()
        sort_function(data_copy)
        end_time = time.perf_counter()
        times.append(end_time - start_time)
    
    return times, algorithm


def calculate_statistics(times: List[float]) -> Dict[str, float]:
    """Calculate statistical measures for timing results"""
    return {
        'average_time': statistics.mean(times),
        'min_time': min(times),
        'max_time': max(times),
        'std_deviation': statistics.stdev(times) if len(times) > 1 else 0.0
    }


def process_algorithm(algorithm: str, sort_function: Callable[[List[int]], None], 
                     data: List[int], num_runs: int, results: List[Dict]) -> None:
    """Process a single algorithm with multiple runs"""
    times, algo_name = run_sort_multiple(algorithm, sort_function, data, num_runs)
    stats = calculate_statistics(times)
    
    result = {
        'algorithm': algo_name,
        'runs': num_runs,
        'times': times,
        **stats
    }
    
    with results_lock:
        results.append(result)


def main():
    """Main function"""
    parser = argparse.ArgumentParser(
        description='Sorting algorithms performance measurement',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog='''
Examples:
  %(prog)s --file data.txt --algorithms quick_sort,merge_sort
  %(prog)s --file data.txt --algorithms bubble_sort --runs 15
  %(prog)s --file data.txt --algorithms quick_sort,merge_sort,heap_sort --runs 5
        '''
    )
    
    parser.add_argument('--file', required=True, help='Input file containing integers')
    parser.add_argument('--algorithms', required=True, 
                       help='Comma-separated list of algorithms to run')
    parser.add_argument('--runs', type=int, default=10, 
                       help='Number of runs for each algorithm (default: 10)')
    
    args = parser.parse_args()
    
    # Validate runs parameter
    if args.runs < 1:
        print("Error: Number of runs must be at least 1.", file=sys.stderr)
        return 1
    
    # Available algorithms
    algorithms = {
        'bubble_sort': bubble_sort,
        'selection_sort': selection_sort,
        'insertion_sort': insertion_sort,
        'quick_sort': quick_sort,
        'merge_sort': merge_sort,
        'heap_sort': heap_sort,
        'counting_sort': counting_sort,
        'radix_sort': radix_sort,
        'bucket_sort': bucket_sort
    }
    
    # Parse chosen algorithms
    chosen_algorithms = [algo.strip() for algo in args.algorithms.split(',')]
    
    # Validate algorithms
    for algorithm in chosen_algorithms:
        if algorithm not in algorithms:
            print(f"Error: Unknown algorithm: {algorithm}", file=sys.stderr)
            print(f"Available algorithms: {', '.join(algorithms.keys())}", file=sys.stderr)
            return 1
    
    # Read data
    try:
        data = read_file(args.file)
    except (FileNotFoundError, ValueError) as e:
        print(f"Error reading file: {e}", file=sys.stderr)
        return 1
    
    if not data:
        print("Error: No data to sort.", file=sys.stderr)
        return 1
    
    # Run algorithms concurrently
    results = []
    
    with ThreadPoolExecutor(max_workers=len(chosen_algorithms)) as executor:
        futures = []
        
        for algorithm in chosen_algorithms:
            sort_function = algorithms[algorithm]
            future = executor.submit(
                process_algorithm, algorithm, sort_function, data, args.runs, results
            )
            futures.append(future)
        
        # Wait for all threads to complete
        for future in futures:
            future.result()
    
    # Sort results by algorithm name for consistent output
    results.sort(key=lambda x: x['algorithm'])
    
    # Create results directory
    results_dir = "../../resources/results/"
    results_file = os.path.join(results_dir, "results_python.json")
    
    os.makedirs(results_dir, exist_ok=True)
    
    # Write results to file
    try:
        with open(results_file, 'w') as outfile:
            json.dump(results, outfile, indent=4)
        
        # Print results to console
        print(json.dumps(results, indent=4))
        print(f"Sorting completed. Results saved to {results_file}")
        
    except IOError as e:
        print(f"Error: Could not create results file at {results_file}: {e}", file=sys.stderr)
        # Fallback to current directory
        fallback_file = "results_python.json"
        try:
            with open(fallback_file, 'w') as outfile:
                json.dump(results, outfile, indent=4)
            print(json.dumps(results, indent=4))
            print(f"Sorting completed. Results saved to {fallback_file}")
        except IOError as e2:
            print(f"Error: Could not write results file: {e2}", file=sys.stderr)
            return 1
    
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: algorithms/cpp/algorithms.cpp
// This file is part of the C++ Algorithms project.

// Include necessary headers
#include "nlohmann/json.hpp"

#include <functional>
#include <algorithm>
#include <iostream>
#include <fstream>
#include <numeric>
#include <cstdlib>
#include <vector>
#include <string>
#include <thread>
#include <chrono>
#include <mutex>
#include <queue>
#include <cmath>
#include <map>

using json = nlohmann::json;
std::mutex output_mutex;


// --- Sorting algorithms ---

// Bubble sort
void bubble_sort(std::vector<int>& array) {
    size_t size = array.size();
    bool swapped;

    if (size < 2)
        return;

    do {
        swapped = false;

        for (size_t index = 1; index < size; ++index)
            if (array[index - 1] > array[index]) {
                std::swap(array[index - 1], array[index]);
                swapped = true;
            }

        --size;
    } while (swapped);

    return;
}

// Selection sort
void selection_sort(std::vector<int>& array) {
    size_t size = array.size();

    if (size < 2)
        return;

    for (size_t index = 0; index < size - 1; ++index) {
        size_t min_index = index;

        for (size_t j = index + 1; j < size; ++j)
            if (array[j] < array[min_index])
                min_index = j;

        std::swap(array[index], array[min_index]);
    }

    return;
}

// Insertion sort
void insertion_sort(std::vector<int>& array) {
    size_t size = array.size();

    if (size < 2)
        return;

    for (size_t index = 1; index < size; ++index) {
        int key = array[index];
        size_t j = index - 1;

        while (j < size && array[j] > key) {
            array[j + 1] = array[j];
            --j;
        }

        array[j + 1] = key;
    }

    return;
}

// Quick sort
int partition(std::vector<int>& array, int low, int high) {
    int pivot = array[high];
    int i = low - 1;

    for (int j = low; j < high; ++j)
        if (array[j] < pivot) {
            ++i;
            std::swap(array[i], array[j]);
        }

    std::swap(array[i + 1], array[high]);

    return i + 1;
}

void quick_sort_recursive(std::vector<int>& array, int low, int high) {
    if (low < high) {
        int pi = partition(array, low, high);

        quick_sort_recursive(array, low, pi - 1);
        quick_sort_recursive(array, pi + 1, high);
    }

    return;
}

void quick_sort(std::vector<int>& array) {
    if (array.size() < 2)
        return;

    quick_sort_recursive(array, 0, array.size() - 1);

    return;
}

// Merge sort
void merge(std::vector<int>& array, int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    std::vector<int> L(n1), R(n2);

    for (int i = 0; i < n1; ++i)
        L[i] = array[left + i];
    
    for (int j = 0; j < n2; ++j)
        R[j] = array[mid + 1 + j];

    int i = 0;
    int j = 0;
    int k = left;

    while (i < n1 && j < n2)
        array[k++] = (L[i] <= R[j]) ? L[i++] : R[j++];

    while (i < n1)
        array[k++] = L[i++];

    while (j < n2)
        array[k++] = R[j++];

    return;
}

void merge_sort_recursive(std::vector<int>& array, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;

        merge_sort_recursive(array, left, mid);
        merge_sort_recursive(array, mid + 1, right);
        merge(array, left, mid, right);
    }

    return;
}

void merge_sort(std::vector<int>& array) {
    if (array.size() < 2)
        return;

    merge_sort_recursive(array, 0, array.size() - 1);

    return;
}

// Heap sort
void heapify(std::vector<int>& array, int n, int i) {
    int largest = i;
    int left = (2 * i) + 1;
    int right = (2 * i) + 2;

    if (left < n && array[left] > array[largest])
        largest = left;

    if (right < n && array[right] > array[largest])
        largest = right;

    if (largest != i) {
        std::swap(array[i], array[largest]);
        heapify(array, n, largest);
    }

    return;
}

void heap_sort(std::vector<int>& array) {
    size_t size = array.size();

    if (size < 2)
        return;

    for (size_t i = static_cast<size_t>(size / 2) - 1; i != static_cast<size_t>(-1); --i)
        heapify(array, size, i);

    for (size_t i = size - 1; i > 0; --i) {
        std::swap(array[0], array[i]);
        heapify(array, i, 0);
    }

    return;
}

// Counting sort
void counting_sort(std::vector<int>& array) {
    if (array.empty())
        return;

    if (array.size() < 2)
        return;

    int max_value = *std::max_element(array.begin(), array.end());
    int min_value = *std::min_element(array.begin(), array.end());
    int range = max_value - min_value + 1;
    std::vector<int> count(range, 0);

    for (int num : array)
        ++count[num - min_value];

    int index = 0;

    for (int i = 0; i < range; ++i)
        while (count[i]--)
            array[index++] = i + min_value;

    return;
}

// Radix sort
void radix_sort(std::vector<int>& array) {
    if (array.empty())
        return;

    if (array.size() < 2)
        return;

    int max_value = *std::max_element(array.begin(), array.end());
    int exp = 1;

    while (max_value / exp > 0) {
        std::vector<int> output(array.size());
        int count[10] = {0};

        for (int num : array)
            ++count[(num / exp) % 10];

        for (int i = 1; i < 10; ++i)
            count[i] += count[i - 1];

        for (int i = static_cast<int>(array.size()) - 1; i >= 0; --i) {
            output[count[(array[i] / exp) % 10] - 1] = array[i];
            --count[(array[i] / exp) % 10];
        }

        array = output;
        exp *= 10;
    }

    return;
}

// Bucket sort
void bucket_sort(std::vector<int>& array) {
    if (array.empty())
        return;

    if (array.size() < 2)
        return;

    int max_value = *std::max_element(array.begin(), array.end());
    size_t bucket_count = static_cast<size_t>(std::sqrt(array.size()));
    std::vector<std::vector<int>> buckets(bucket_count);

    for (int num : array) {
        size_t bucket_index = static_cast<size_t>((static_cast<double>(num) / (max_value + 1)) * bucket_count);
        if (bucket_index >= bucket_count)
            bucket_index = bucket_count - 1;
        buckets[bucket_index].push_back(num);
    }

    array.clear();

    for (auto& bucket : buckets) {
        if (bucket.size() < 2) {
            array.insert(array.end(), bucket.begin(), bucket.end());
            continue;
        }

        std::sort(bucket.begin(), bucket.end());
        array.insert(array.end(), bucket.begin(), bucket.end());
    }

    return;
}


// --- Utility functions ---

std::vector<int> read_file(const std::string& path) {
    std::ifstream infile(path);

    if (!infile)
        throw std::runtime_error("Could not open file: " + path);

    int number;    
    std::vector<int> data;

    while (infile >> number)
        data.push_back(number);

    infile.close();
    return data;
}

std::pair<std::vector<double>, std::string> run_sort_multiple(const std::string& algorithm,
                                                              const std::function<void(std::vector<int>&)>& sort_function,
                                                              const std::vector<int>& data,
                                                              int runs = 10) {
    std::vector<double> times;
    times.reserve(runs);
    
    for (int i = 0; i < runs; ++i) {
        auto copy = data;
        auto start = std::chrono::high_resolution_clock::now();
        sort_function(copy);
        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> duration = end - start;
        times.push_back(duration.count());
    }
    
    return {times, algorithm};
}

// --- Main function ---

int main(int argc, char* argv[]) {
    std::string file_path;
    std::vector<std::string> chosen_algorithms;
    int num_runs = 10; // Default number of runs

    // Check for help first
    if (argc == 1 || (argc == 2 && (std::string(argv[1]) == "--help" || std::string(argv[1]) == "-h"))) {
        std::cout << "C++ Sorting Algorithms Benchmark Tool\n\n";
        std::cout << "Usage: " << argv[0] << " --file <input_file> --algorithms <algorithm_list> [--runs <number>]\n\n";
        std::cout << "Required Arguments:\n";
        std::cout << "  --file <path>           Input file containing space-separated integers\n";
        std::cout << "  --algorithms <list>     Comma-separated list of algorithms to run\n\n";
        std::cout << "Optional Arguments:\n";
        std::cout << "  --runs <number>         Number of runs per algorithm (default: 10)\n";
        std::cout << "  --help, -h              Show this help message\n\n";
        std::cout << "Available Algorithms:\n";
        std::cout << "  bubble_sort             Bubble Sort (O(n^2))\n";
        std::cout << "  selection_sort          Selection Sort (O(n^2))\n";
        std::cout << "  insertion_sort          Insertion Sort (O(n^2))\n";
        std::cout << "  quick_sort              Quick Sort (O(n log n) average)\n";
        std::cout << "  merge_sort              Merge Sort (O(n log n))\n";
        std::cout << "  heap_sort               Heap Sort (O(n log n))\n";
        std::cout << "  counting_sort           Counting Sort (O(n + k))\n";
        std::cout << "  radix_sort              Radix Sort (O(d x (n + k)))\n";
        std::cout << "  bucket_sort             Bucket Sort (O(n + k))\n\n";
        std::cout << "Examples:\n";
        std::cout << "  " << argv[0] << " --file data.txt --algorithms quick_sort,merge_sort\n";
        std::cout << "  " << argv[0] << " --file data.txt --algorithms bubble_sort --runs 15\n";
        std::cout << "  " << argv[0] << " --file data.txt --algorithms quick_sort,merge_sort,heap_sort --runs 5\n\n";
        std::cout << "Output:\n";
        std::cout << "  Results are saved to: ../../resources/results/results_cpp.json\n";
        return 0;
    }

    for (int index = 1; index < argc; ++index) {
        std::string arg = argv[index];

        if (arg == "--file" && index + 1 < argc)
            file_path = argv[++index];
        else if (arg == "--algorithms" && index + 1 < argc) {
            std::string algorithms = argv[++index];
            size_t pos = 0;

            while ((pos = algorithms.find(',')) != std::string::npos) {
                chosen_algorithms.push_back(algorithms.substr(0, pos));
                algorithms.erase(0, pos + 1);
            }

            chosen_algorithms.push_back(algorithms);
        }
        else if (arg == "--runs" && index + 1 < argc) {
            num_runs = std::stoi(argv[++index]);
            if (num_runs < 1) {
                std::cerr << "Number of runs must be at least 1." << std::endl;
                return 1;
            }
        }
        else if (arg == "--help" || arg == "-h") {
            // Help already handled above, but include here for completeness
            return 0;
        }
        else {
            std::cerr << "Unknown argument: " << arg << std::endl;
            std::cerr << "Use --help for usage information." << std::endl;
            return 1;
        }
    }

    if (file_path.empty() || chosen_algorithms.empty()) {
        std::cerr << "Error: Missing required arguments." << std::endl;
        std::cerr << "Usage: " << argv[0] << " --file <path> --algorithms <alg1,alg2,...> [--runs <number>]" << std::endl;
        std::cerr << "Use --help for detailed usage information." << std::endl;
        return 1;
    }

    std::map<std::string, std::function<void(std::vector<int>&)>> algorithms = {
        {"bubble_sort", bubble_sort},
        {"selection_sort", selection_sort},
        {"insertion_sort", insertion_sort},
        {"quick_sort", quick_sort},
        {"merge_sort", merge_sort},
        {"heap_sort", heap_sort},
        {"counting_sort", counting_sort},
        {"radix_sort", radix_sort},
        {"bucket_sort", bucket_sort}
    };

    std::vector<int> data;
    try {
        data = read_file(file_path);
    } 
    catch (const std::runtime_error& e) {
        std::cerr << "Error reading file: " << e.what() << std::endl;
        return 1;
    }

    if (data.empty()) {
        std::cerr << "No data to sort." << std::endl;
        return 1;
    }

    json results = json::array();
    std::vector<std::thread> threads;

    for (const auto& algorithm : chosen_algorithms) {
        auto it = algorithms.find(algorithm);
        if (it != algorithms.end()) {
            threads.emplace_back([&, algorithm, sort_function = it->second, num_runs]() {
                auto [times, algo_name] = run_sort_multiple(algorithm, sort_function, data, num_runs);
                
                // Calculate statistics
                double avg_time = std::accumulate(times.begin(), times.end(), 0.0) / times.size();
                double min_time = *std::min_element(times.begin(), times.end());
                double max_time = *std::max_element(times.begin(), times.end());

                // Calculate standard deviation
                double sum_sq = 0.0;
                for (double time : times)
                    sum_sq += (time - avg_time) * (time - avg_time);
                double std_dev = std::sqrt(sum_sq / times.size());

                std::lock_guard<std::mutex> lock(output_mutex);
                results.push_back({
                    {"algorithm", algo_name},
                    {"runs", num_runs},
                    {"times", times},
                    {"average_time", avg_time},
                    {"min_time", min_time},
                    {"max_time", max_time},
                    {"std_deviation", std_dev}
                });
            });
        }
        else {
            std::cerr << "Unknown algorithm: " << algorithm << std::endl;
            return 1;
        }
    }

    for (auto& thread : threads)
        if (thread.joinable())
            thread.join();

    // Create the results directory path relative to the project root
    std::string results_dir = "../../resources/results/";
    std::string results_file = results_dir + "results_cpp.json";
    
    // Create directory if it doesn't exist (using system call)
    system("mkdir -p ../../resources/results/");

    std::ofstream outfile(results_file);
    if (!outfile) {
        std::cerr << "Error: Could not create results file at " << results_file << std::endl;
        // Fallback to current directory
        results_file = "results_cpp.json";
        outfile.open(results_file);
    }
    
    std::cout << results.dump(4) << std::endl;
    outfile << results.dump(4);
    outfile.close();
    std::cout << "Sorting completed. Results saved to " << results_file << std::endl;

    return 0;
}
